### PR TITLE
Feat/#454 implement ErrorOutOfGasCreate error state

### LINKED ISSUE #454: Implement `ErrorOutOfGasCREATE2` error state
For `CREATE2` opcodes which may run out of gas.

Scroll team implementation: https://github.com/scroll-tech/zkevm-circuits/pull/507

### PR DESCRIPTION
**Pull request by KimiWu123**
closed #454 and #459

---

[EIP-3860](https://eips.ethereum.org/EIPS/eip-3860#rules) includes
- limit the max size of init code to `49152` 
- add extra gas required for init code, `2 * len(ceil(init_code/32))`

This PR includes
1. move `CREATE/2` oog from `error_oog_dynamic_memory_expansion.py` to `error_oog_create.py`
2. support EIP-3860
   - `create.py` add  extra gas cost for init code. 
   - `begin_tx.py` add  extra gas cost for init code
   - `error_oog_create.py` handles errors of oog and max size of init code  (see [Rules 1 and 3](https://eips.ethereum.org/EIPS/eip-3860#rules)).    
   - `error_gas_uint_overflow.py` handles extra gas required for init code

### CODE REVIEW

File: src/zkevm_specs/evm_circuit/execution/error_gas_uint_overflow.py [hunk 1]
```diff
@@ -32,7 +36,7 @@ def error_gas_uint_overflow(instruction: Instruction):
         is_log3,
         is_log4,
         is_sha3,
-        is_exp,
+        _,
```
Comment (ChihChengLiang) [inline src/zkevm_specs/evm_circuit/execution/error_gas_uint_overflow.py:39]: is_exp has dynamic gas of 50 * exponent_byte_size
I think we should put is_exp in the `is_dynamic_gas` variable below

File: src/zkevm_specs/evm_circuit/execution/error_gas_uint_overflow.py [hunk 2]
```diff
@@ -17,7 +21,7 @@ def error_gas_uint_overflow(instruction: Instruction):
 
     (
         is_call,
-        is_callcode,
+        _,
```
Comment (ChihChengLiang) [inline src/zkevm_specs/evm_circuit/execution/error_gas_uint_overflow.py:24]: We should put is_callcode in is_dynamic_gas too.

**Review by ChihChengLiang — APPROVED** [review]
LGTM after the feedback is addressed.

File: src/zkevm_specs/evm_circuit/execution/error_gas_uint_overflow.py (hunk 2 above)
Comment (KimiWu123) [reply src/zkevm_specs/evm_circuit/execution/error_gas_uint_overflow.py]: fixed in [567055f](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/481/commits/567055f217dc65cb2e9d3c020db60f97f1ecc20e)

File: src/zkevm_specs/evm_circuit/execution/error_gas_uint_overflow.py (hunk 1 above)
Comment (KimiWu123) [reply src/zkevm_specs/evm_circuit/execution/error_gas_uint_overflow.py]: I found that it only takes care memory expansion cost which EXP opcode doesn't have. So, I just remove EXP in `ErrorGasUintOverflow` gadget. We might need to have more investigation on `ErrorGasUintOverflow` error state and we could do it when we apply Scroll's approch (performing `ErrorGasUintOverflow` check in each oog case)

File: specs/error_state/ErrorOutOfGasCreate.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+### Constraints
+
+1. Current opcode is `CREATE` or `CREATE2`.
+2. At least one of below conditions is met:
+  - `gas_left < gas_cost`.
+  - `len(init_code) > MAX_INIT_CODE_SIZE`
+3. Current call fails.
+4. Common error constraints: 
+  - Current call fails.
```
[the remark this answers is not in the document]
Comment (KimiWu123) [reply specs/error_state/ErrorOutOfGasCreate.md]: fixed in[6a5ebce](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/481/commits/6a5ebce3d0547cad0a575c84ba53a15a581de290)

File: specs/opcode/F0CREATE_F5CREATE2.md [hunk 4]
```diff
@@ -87,11 +90,11 @@ The circuit takes the starting `rw_counter` as next call's `call_id` to make sur
...
 
-It then checks that `callee.is_persistent = caller.is_persistent and caller.is_success`.
-If the caller is not persistent, we need to propagate the `rw_counter_end_of_reversion` to make sure every state update in the new call has a corresponding reversion.
-
-It stores the current call context by writing the current values `rw_table` and checks that the new call context is setup correctly by reading to `rw_table`, and then does a step state transition to the call and begins the execution o the initialization bytecode.
-
+If all the pre-checks pass, then
+  - It checks that `callee.is_persistent = caller.is_persistent and caller.is_success`.
+  - If the caller is not persistent, we need to propagate the `rw_counter_end_of_reversion` to make sure every state update in the new call has a corresponding reversion.
+  -  It stores the current call context by writing the current values `rw_table` and checks that the new call context is setup correctly by reading to `rw_table`, and then does a step state transition to the call and begins the execution o the initialization bytecode.
```
[the remark this answers is not in the document]
Comment (KimiWu123) [reply specs/opcode/F0CREATE_F5CREATE2.md]: fixed in[6a5ebce](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/481/commits/6a5ebce3d0547cad0a575c84ba53a15a581de290)

File: specs/opcode/F0CREATE_F5CREATE2.md [hunk 5]
```diff
@@ -14,14 +14,14 @@ The following values are popped from the stack:
...
-If the initialization call is successful, the new address will be pushed to the stack.
-If not, 0 will be pushed instead.
+If the initialization call is successful, the new address will be pushed to the stack. If not, 0 will be pushed instead.
 
 The gadget does several things in the caller's call context:
-1. Expands memory
-2. Add `new_address` to the access list
-3. Calculate gas cost for this step.
-4. Uses EIP150 to calculate how much gas will remain in the caller's context.
+1. Do following pre-checks, stack depth, sender's balance, sender's nonce and contract address existence. If any one of them fails, restores the caller's context.
```
[the remark this answers is not in the document]
Comment (KimiWu123) [reply specs/opcode/F0CREATE_F5CREATE2.md]: fixed in[6a5ebce](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/481/commits/6a5ebce3d0547cad0a575c84ba53a15a581de290)

File: specs/error_state/ErrorOutOfGasDynamicMemoryExpansion.md [hunk 6]
```diff
@@ -2,28 +2,16 @@
...
-else:
-    constant_gas = 0
-```
 
-The dynamic gas is calculated as:
-
-```
+constant_gas = 0
 dynamic_gas = memory_expansion_cost
 ```
```
[the remark this answers is not in the document]
Comment (KimiWu123) [reply specs/error_state/ErrorOutOfGasDynamicMemoryExpansion.md]: fixed in [0d85670](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/481/commits/0d8567045b3638c47c2ba3c874f67274b7846deb)

**Review by rrtoledo — APPROVED** [review]
LGTM :)

**Pull request merged**